plugins/auth-backend/src/providers/oauth2-proxy/provider.ts

```diff
@@ -0,0 +1,195 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import express from 'express';
+import { Logger } from 'winston';
+import { AuthenticationError } from '@backstage/errors';
+import {
+  AuthHandler,
+  SignInResolver,
+  AuthProviderFactory,
+  AuthProviderRouteHandlers,
+  AuthResponse,
+} from '../types';
+import { CatalogIdentityClient } from '../../lib/catalog';
+import { JWT } from 'jose';
+import { IdentityClient } from '../../identity';
+import { TokenIssuer } from '../../identity/types';
+import { prepareBackstageIdentityResponse } from '..';
+
+export const OAUTH2_PROXY_JWT_HEADER = 'X-OAUTH2-PROXY-ID-TOKEN';
+
+/**
+ * @public
+ */
+export type OAuth2ProxyResult<JWTPayload> = {
+  fullProfile: JWTPayload;
+  expiresInSeconds?: number;
+  accessToken: string;
+};
+
+/**
+ * @public
+ */
+export type Oauth2ProxyProviderOptions<JWTPayload> = {
+  /**
+   * Configure an auth handler to generate a profile for the user.
+   */
+  authHandler: AuthHandler<OAuth2ProxyResult<JWTPayload>>;
+
+  /**
+   * Configure sign-in for this provider, without it the provider can not be used to sign users in.
+   */
+  signIn: {
+    /**
+     * Maps an auth result to a Backstage identity for the user.
+     */
+    resolver: SignInResolver<OAuth2ProxyResult<JWTPayload>>;
+  };
+};
+
+interface Options<JWTPayload> {
+  logger: Logger;
+  signInResolver: SignInResolver<OAuth2ProxyResult<JWTPayload>>;
+  authHandler: AuthHandler<OAuth2ProxyResult<JWTPayload>>;
+  tokenIssuer: TokenIssuer;
+  catalogIdentityClient: CatalogIdentityClient;
+}
+
+/**
+ * @public
+ */
+export class Oauth2ProxyAuthProvider<JWTPayload>
+  implements AuthProviderRouteHandlers
+{
+  private readonly logger: Logger;
+  private readonly catalogIdentityClient: CatalogIdentityClient;
+  private readonly signInResolver: SignInResolver<
+    OAuth2ProxyResult<JWTPayload>
+  >;
+  private readonly authHandler: AuthHandler<OAuth2ProxyResult<JWTPayload>>;
+  private readonly tokenIssuer: TokenIssuer;
+
+  constructor(options: Options<JWTPayload>) {
+    this.catalogIdentityClient = options.catalogIdentityClient;
+    this.logger = options.logger;
+    this.tokenIssuer = options.tokenIssuer;
+    this.signInResolver = options.signInResolver;
+    this.authHandler = options.authHandler;
+  }
+
+  frameHandler(): Promise<void> {
+    return Promise.resolve(undefined);
+  }
+
+  async refresh(req: express.Request, res: express.Response): Promise<void> {
+    try {
+      const result = this.getResult(req);
+
+      const response = await this.handleResult(result);
+
+      res.status(200);
```

Comment:
No need to set an explicit 200 response since it's default

Reply:
Adjusted 👍